### DISCUSSION TITLE
Proposal: Supply Optional Timeout to lock Statement

### DISCUSSION THREAD

**Discussion by jtwine** [dotnet/csharplang#1450] [Language Ideas]
First proposal, so please bear with me if I do something wrong.  I have been having to deal with a few surprise deadlock situations recently and have converted some code to make manual use of the `Monitor` class to be able to specify timeouts via `Monitor.TryEnter(...)`.  

It dawned on me that it would be a nice feature to be able to specify the timeout on the `lock` statement, and have it throw an exception if the timeout expires without entering the lock:
```
try
{
    lock( m_someLockObject, 500 )
    {
        //
        // Do Something In The CS...
        //
    }
}
catch( TimeoutException tEx )
{
    //
    // Handle Timeout/Deadlock Situation...
    //
} 
```
Or, even better, have this new version return a `bool` indicating if the lock was entered or not, to save the overhead of exceptions and allow the error to be handled in a more natural way?
```
if( lock( m_someLockObject, 500 ) )
{
    //
    // Do Something In The CS...
    //
}
else
{
    //
    // Handle Timeout/Deadlock Situation...
    //
}
```
The scope of the `lock` would _borrow_ the scope of the `if` statement in this example.

I did a search in the existing issues using the word "lock" and did not find anything like this in the 3 pages of results, so I figured no harm in asking.  Yes, I know the same can be achieved with manual use of the `Monitor` object, but I believe this would be a little easier to write, easier to understand, and does not really hide what is going on from the developer.  Just as we use the `lock` statement today instead of `Monitor`, we could use something like the above examples so that we are not using `lock` in one part of the code and `Monitor.TryLock(...)` in other parts of the code, which can be confusing to less-experienced developers.

Thanks for reading!

Peace!

-=- James.

**Comment by bbarry** [comment]
There is no reason to rely on an exception here, `TryEnter` provides a bool for that purpose. I don't like `if( lock( m_someLockObject, 500 ) )` because lock is a statement not an expression (you couldn't for example combine some bool logic there), but for the one time I think I've ever used `TryEnter` it would be nice to have a way to avoid the boilerplate:

```csharp
var lockObj = _lockobject;
bool lockTaken = false;
try
{
    Monitor.TryEnter(lockObj, timeout, ref lockTaken);
    if (lockTaken)
    {
        // The critical section.
    }
    else
    {
        // The lock was not acquired.
    }
}
finally
{
    if (lockTaken)
    {
        Monitor.Exit(lockObj);
    }   
}
```

vs

```csharp
lock(_lockojbect, timeout)
{
    // The critical section.
}
fail
{
    // The lock was not acquired.
}
```

I don't think you could simply reuse else either due to the ambiguity but perhaps a new contextual keyword could be used instead.

Reviewing code in CoreFx, there are a few patterns using `Monitor.TryEnter` (`TryEnter` without a timeout is the same as `TryEnter` with a timeout of `0`):

* in [VolatileEnlistmentMultiplexing](https://github.com/dotnet/corefx/blob/6abde68de9611895fa8697fb7fc17105e906081e/src/System.Transactions.Local/src/System/Transactions/VolatileEnlistmentMultiplexing.cs#L79) and [DbConnectionPool](https://github.com/dotnet/corefx/blob/a3c2df2945ed26c42502faa8c4a47c5ccc9e1630/src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionPool.cs#L1491) the basic pattern is used
* in [TdsParserStateObject](https://github.com/dotnet/corefx/blob/4c1704fec0cebfd05a9c099139c6c0217722d31c/src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs#L585),  [DbReferenceCollection](https://github.com/dotnet/corefx/blob/9dc5aad6c1db08049f00d01f504a742cc9751199/src/Common/src/System/Data/ProviderBase/DbReferenceCollection.cs#L140) and [XmlSchemaSet](https://github.com/dotnet/corefx/blob/40368b2aa9ade8e5f495bfb6f474cd16561abe59/src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaSet.cs#L390) various forms of loops are done to obtain locks without explicitly deadlocking
* in [SqlInternalConnectionTds](https://github.com/dotnet/corefx/blob/86f5fb24b56b2333a9542bb6d80db0128a0b119e/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs#L211) something completely different is being done

In total this proposal would simplify 2 existing methods in CoreFx (nothing found in CoreCLR or Roslyn).

**Comment by bondsbw** [comment]
Could you add specifics for how this would work?  E.g. code or pseudo-code representing exactly what occurs?

_Edit: @bharry beat me to it, does that seem to explain it?_

**Comment by jtwine** [comment]
The overall goal would be to avoid the additional coding (`bool` variable and the `try`/`finally`) required around using `Monitor.TryEnter(...)` directly as @bbarry has indicated.  The `lock`/`fail` method they indicated would also solve the problem nicely.  

Remember that I said that I could do everything I needed to do by using `Monitor` directly.  The point was some additional _sugar_ to make that easier, just like `lock` does today.  

Peace!

-=- James.

**Comment by scalablecory** [comment]
I don't think this would have enough impact to consider adding it to the language.

**Comment by jtwine** [comment]
@scalablecory Because?  Have you not had a lot of experience having to deal with deadlocks and situations where using critical sections with timeouts is helpful?

**Comment by CyrusNajmabadi** [comment]
deadlocks only convey a logic bug to me.  So, given a deadlock, i don't think a timeout is a good solution.  Instead, i would prefer to just refactor things to make the deadlock impossible.

**Comment by CyrusNajmabadi** [comment]
Note: roslyn is heavily concurrent, with tons of threads and async work happening at any time across multiple processes.  Despite that, i haven't ever once wanted a lock with a timeout.   And even if there was a case that i felt warranted it... i would be ok just writing the code manually.

For me, this should be a language feature if it was hugely common to want to provide this. i.e. at least 5% of all lock taking was done with a timeout.

**Comment by jtwine** [comment]
Deadlocks are not the only thing that can benefit from this sort of thing, which BTW, is useful to help detect the deadlock and report/log on it, instead of having a system just grind to a halt for no apparent reason.  And, it may be better to keep running if/where possible, especially if what you are doing is more than some little app that can simply be retried/restarted if it hangs up.  If devstudio hangs up (again) I can restart it - no big deal.  Not the case with some larger software systems that deal with multiple clients (better to detect the deadlock and abort one client's action than to hang up for all clients), so comparisons between the two would be inappropriate.

Timing out on a lock is also useful in situations where you are, for example, dealing with rapidly incoming data from a real-time sensor, receiving DMA data, etc.  If you have a producer/consumer setup with multiple threads and for some reason the producer gets delayed on a lock for too long, you might want to detect that and drop the data because it might be stale by the time the lock releases and gets to the consumer.  Of course, this applies to situations where the output needs to be as close to real-time as possible, and not everyone has experience with that kind of system (and they need to appreciate their own limitations in that case).

Of course, I am OK with writing out the long form as is required right now.  But I cannot be the only developer that has to do this on a regular basis, so I figured it would be useful.

Lastly, it really is inappropriate for those that are *not* using the long form version regularly to say that the feature is not needed - they lack the experience with it to make such a statement.  For example, it is inappropriate for me to state that `sealed` should be deprecated, just because *I* rarely use it.

(Thanks to@jnm2 for the spell-check on 'deprecated'!)

**Comment by jnm2** [comment]
(I apologize) I see a lot of people saying ‘depreciated’ when they mean ‘deprecated.’

**Comment by CyrusNajmabadi** [comment]
>Of course, I am OK with writing out the long form as is required right now. But I cannot be the only developer that has to do this on a regular basis, so I figured it would be useful.

It def could be useful :)

My only point was that is seems very niche given the ecosystem.  We'll see if more requests come in for this sort of thing.

**Comment by orthoxerox** [comment]
@jnm2 I apprecate your comment 😀

**Comment by AdamFrisby** [comment]
Just bumping this one - I wrote up a very similar ticket over here with a few more details: https://github.com/dotnet/runtime/issues/55807 - every few years I run into a case where this would be a substantial time and headache saver. While not necessarily a commonly requested feature, I cannot overstate the utility value when you do need it.

**Reply by En3Tho** [reply]
Utility value means saving 5 lines of code once a decade?

**Reply by AdamFrisby** [reply]
Far more. Saving days of work and a massive headache tracking hard to debug deadlocks. 

In codebases which use lock() extensively, it would represent a significant number of LOC.

I recall a few years ago, a coworker made a reflection tool which used Reflection.Emit to rewrite this in existing assemblies to locate bugs like this; someone going to that level of work to write a new debugging utility beyond the built-in debugger would imply something that may want to be incorporated.

**Reply by En3Tho** [reply]
That doesn't really save the code from being error-prone. It's basically like try catch with empty catch block. Also, what is keeping you from writing a simple function Lock(obj, timeout, action/func)  that basically does the same thing at cost of one little allocation? You don't need to wait years for a feature with super low chance of being implemented.

**Reply by AdamFrisby** [reply]
No, I'm not waiting on the request - I'm merely pointing out it has utility.

This is the opposite of a empty try/catch block - this surfaces errors rather than suppresses them. By invoking an exception (see the proposal I linked above) you create a DeadlockException that can help improve code by surfacing otherwise hidden errors.
